### PR TITLE
Laboratory work #2, Ivan Shmakov - 23FPL1

### PR DESCRIPTION
**Pull request by CloudWhisp**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -18,7 +19,21 @@ def tokenize(text: str) -> list[str] | None:
...
+        if not sign.isalpha():
+            added = ' '
+        else:
+            added = sign
+        new_text = f'{new_text}{added}'
+    while '  ' in new_text:
+        new_text = new_text.replace('  ', ' ')
+    for token in new_text.split():
+        tokenized_list.append(token)
+    return tokenized_list
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:36]: this is very difficult to understand, simplify

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -33,7 +48,21 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+        if not isinstance(data, str):
+            return None
+    for data in stopwords:
+        if not isinstance(data, str):
+            return None
+    tokens_cleared = []
+    for elem in tokens:
+        if not elem in stopwords:
+            tokens_cleared.append(elem)
+    return tokens_cleared
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:65]: use list comprehension

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -47,7 +76,18 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+    unique_words = []
+    for text in documents:
+        if not isinstance(text, list):
+            return None
+        for word in text:
+            if not isinstance(word, str):
+                return None
+            if not word in unique_words:
+                unique_words.append(word)
+    return unique_words
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:90]: try to use list comprehension

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -77,8 +132,25 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+        doc_number = 0
+        for doc in documents:
+            if not isinstance(doc, list):
+                return None
+            for word in doc:
+                if not isinstance(word, str):
+                    return None
+            if elem in doc:
+                doc_number += 1
+                continue
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:151]: is it really need to use `continue` here

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Good work! Now fix comments

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (CloudWhisp) [reply lab_2_retrieval_w_bm25/main.py:151]: Fixed

File: lab_2_retrieval_w_bm25/main.py (hunk 2 above)
Comment (CloudWhisp) [reply lab_2_retrieval_w_bm25/main.py:65]: Done!

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (CloudWhisp) [reply lab_2_retrieval_w_bm25/main.py:82]: Done!

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (CloudWhisp) [reply lab_2_retrieval_w_bm25/main.py]: Done!

**Pull request closed without merging**